x-pack/plugins/security_solution/server/endpoint/services/actions/action_files.ts

```diff
@@ -183,3 +199,119 @@ export const validateActionFileId = async (
+  target_agents: string[];
+  action_id: string;
+}
+
+interface UploadedFile {
+  file: Pick<
+    Required<FileJSON<UploadFileInternalStorageMeta>>,
+    'id' | 'created' | 'updated' | 'name' | 'mimeType' | 'extension' | 'meta' | 'status'
+  > & { size: number; hash: { sha256: string } };
+}
+
+export const createFile = async ({
+  esClient,
+  logger,
+  fileStream,
+  maxFileBytes,
+  agents,
+}: {
+  esClient: ElasticsearchClient;
+  logger: Logger;
+  fileStream: HapiReadableStream;
+  maxFileBytes: number;
+  agents: string[];
+}): Promise<UploadedFile> => {
+  const fileClient = getFileClient(esClient, logger, maxFileBytes);
+
+  const uploadedFile = await fileClient.create<UploadFileInternalStorageMeta>({
+    id: uuidV4(),
+    metadata: {
+      name: fileStream.hapi.filename ?? 'unknown_file_name',
+      mime: fileStream.hapi.headers['content-type'] ?? 'application/octet-stream',
+      meta: {
+        target_agents: agents,
+        action_id: '', // Populated later once we have the action is created
+      },
+    },
+  });
+
+  await uploadedFile.uploadContent(fileStream, undefined, {
+    transforms: [createFileHashTransform()],
+  });
+
+  assert(uploadedFile.data.hash && uploadedFile.data.hash.sha256, 'File hash was not generated!');
+
+  return toUploadedFileInterface(uploadedFile);
+};
+
+const toUploadedFileInterface = (file: File<UploadFileInternalStorageMeta>): UploadedFile => {
+  const { name, created, meta, id, mimeType, size, status, extension, hash, updated } =
+    file.toJSON();
+
+  return {
+    file: {
+      name,
+      created,
+      updated,
+      meta,
+      id,
+      mimeType,
+      status,
+      extension,
+      size: size ?? 0,
+      hash: { sha256: hash?.sha256 ?? '' },
+    },
+  };
+};
+
+/**
+ * Deletes a file by ID
+ * @param esClient
+ * @param logger
+ * @param fileId
+ */
+export const deleteFile = async (
+  esClient: ElasticsearchClient,
+  logger: Logger,
+  fileId: string
+): Promise<void> => {
+  const fileClient = getFileClient(esClient, logger);
+
+  await fileClient.delete({ id: fileId, hasContent: true });
+};
+
+/**
+ * Sets the `meta.action_id` on the file associated with the `upload` action
+ * @param esClient
+ * @param logger
+ * @param action
+ */
+export const setFileActionId = async (
+  esClient: ElasticsearchClient,
+  logger: Logger,
+  action: ActionDetails<ResponseActionUploadOutputContent, ResponseActionUploadParameters>
+): Promise<UploadedFile> => {
+  assert(
+    action.parameters?.file.file_id,
+    `Action [${action.id}] has no 'parameters.file.file_id' defined. Unable to set action id on file record`
+  );
+
+  const fileClient = getFileClient(esClient, logger);
+  const file = await fileClient.get<UploadFileInternalStorageMeta>({
+    id: action.parameters?.file.file_id,
+  });
+
+  await file.update({
+    meta: {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      ...file.data.meta!,
```

Comment:
can we get rid of the bang somehow?

Reply:
Let me see in the next PR is I can somehow narrow the type. The issue here is that `meta` is optional in the `Files` plugin interface, but required in our use case. Will look into it.